### ISSUE TITLE
timeline for the LM support

### ISSUE THREAD

**Issue by xhzhao** [SeanNaren/deepspeech.pytorch#236]
we know that the LM support is WIP now, and do you have any schedule about when this feature will come out? or could i know which branch is related to this work? so that i could track the updates or contribute.

I have another question about the Alternate Decoders in README.

- Can i use the beamsearch without any language model?

- IF yes, how much accuracy benefit could we get?

**Comment by ryanleary** [comment]
It's essentially done and included in master.

Yes you can use beam search without an LM. The accuracy improvement is minimal but I haven't measured recently.

**Comment by SeanNaren** [comment]
LM support is complete, training an LM however is definitely a different story, may make a ticket in case anyone feels like tackling that!

**Comment by xhzhao** [comment]
@ryanleary @SeanNaren thanks for your response, i will build the LM from this [link](http://kheafield.com/code/kenlm/), and try to get a better WER accuracy with beamsearch=300.
BTW, do you have any accuracy data with the kenlm on Librispeech?

**Comment by ryanleary** [comment]
You can just download the LM directly from that link pretrained. I generally see about 3.5% WER reduction (absolute) on the libri clean test set and ~9% WER reduction (absolute) on the libri other test set.

**Comment by xhzhao** [comment]
@ryanleary 3.5%WER reduction is great. How much beamsize do you use? BTW, i downloaded the kenlm and make it sucessfully. but i don't know how to use the language model, i mean i don't know which path should i fill in this parameter:

- lm_path optional binary KenLM language model to use for decoding 

should i use the "test.arpa" path or "test_nounk.arpa" path?

**Comment by xhzhao** [comment]
@ryanleary @SeanNaren I'd like to share some accuracy data of the pretrained model with LM on Librispeech. I use the 3-gram.arpa.gz file from this [link](http://www.openslr.org/11/), and use the pretrained model for librispeech, the accuracy on test-clean looks like this:

| Model        | WER           | CER  |
| ------------- |:-------------:| -----:|
| no LM      | 11.294 | 3.092|
| with LM, Beam10      | 8.13 |   3.174 |
| with LM, beam300 | 6.454|    2.2 |

**Comment by ryanleary** [comment]
Thanks @xhzhao . I'll add it to the release notes. Did you tune the alpha and beta parameters? If not, you might be able do a little better better. Nice result nonetheless.

**Comment by xhzhao** [comment]
@ryanleary i did not tune the alpha or beta, do you know how to tune these 2 parameters?
i add all the accuracy data as follows:
![image](https://user-images.githubusercontent.com/17486215/35713580-0aa8ec3e-0803-11e8-99b6-ffe9e5886532.png)

**Comment by ryanleary** [comment]
If you run `test.py` with `decoder=None` and set `output-path` to a new file, you can output the logits from the acoustic model. Then you can run `tune_decoder.py` however you like and it will run a grid search over the alpha and beta parameters.

Running the `test.py` script and saving the logits to a file let's you easily run multiple decodes in parallel without rerunning the neural net each time.

**Comment by fmobrj** [comment]
Hi, Xiaohui! Great results. Congratulations.

Can you share some ideas about how you got the results showed in SeanNaren’s repo for the deepspeech2 model?

For the base model, without LM decoding, for the val dataset I see you’ve managed to achieve WER ~ 20.  I trained using the default training arguments (No LM, H=800, epochs = 70) and can’t achieve a WER inferior to ~ 27. (python train.py --train-manifest data/libri_train_manifest.csv --val-manifest data/libri_val_manifest.csv --batch-size 32 –cuda)

Did you use the –augment option or set any other parameter to achieve those results? Or did you just run the default command line with the default parameters?

Thank you very much.
